File: bindings/python/insight/signal/estimation.py
"""Estimation functions for signal processing.

Provides the Kalman filter for state estimation.
"""

from insight._insight import signal as _signal

__all__ = [
    "KalmanFilter",
]

#: KalmanFilter class for linear state estimation.
#:
#: Args:
#:     dim_x (int): Dimension of the state vector.
#:     dim_z (int): Dimension of the measurement vector.
#:     dim_u (int): Dimension of the control vector.  Default is 0.
#:     points (int): Number of filter points (ensemble).  Default is 1.
#:     dtype: Data type.  Default is ``float64``.
#:
#: Attributes:
#:     x: State vector.
#:     P: State covariance matrix.
#:     z: Measurement vector.
#:     R: Measurement noise covariance.
#:     Q: Process noise covariance.
#:     F: State transition matrix.
#:     H: Measurement matrix.
#:     dim_x: State dimension (read-only).
#:     dim_z: Measurement dimension (read-only).
#:     points: Number of points (read-only).
#:
#: Methods:
#:     predict(): Predict the next state.
#:     update(z): Update the state with a new measurement.
KalmanFilter = _signal.KalmanFilter

File: bindings/lua/insight/signal/estimation.lua
--- State estimation functions.
-- Provides the KalmanFilter class for linear dynamic system estimation.
-- @module insight.signal.estimation

local native = require("_insight")
local sig = native.signal

local M = {}

local _wrap = require("insight._wrap")

--- Create a KalmanFilter for linear dynamic system state estimation.
-- @int dim_x Number of state variables.
-- @int dim_z Number of measurement inputs.
-- @int[opt=0] dim_u Number of control inputs.
-- @int[opt=1] points Number of filter points.
-- @tparam[opt=ins.float64] DType dtype Data type.
-- @treturn KalmanFilter The constructed filter.
M.KalmanFilter = _wrap({ "dim_x", "dim_z", "dim_u", "points", "dtype" }, function(dim_x, dim_z, dim_u, points, dtype)
  return sig.KalmanFilter_new(dim_x, dim_z, dim_u, points, dtype)
end)

return M
